### ISSUE TITLE
Hierarchical RNN with word and sentence recurrent layers 

### ISSUE THREAD

**Issue by Youmna-H** [fchollet/keras#5516]
I am trying to implement an LSTM that applies a recurrent operation to words in the same sentence independently, then to the sentences in the whole document.
My input shape is accordingly:
sequence = Input(shape=(max_num_sents_in_doc, max_num_words_in_sent, ), dtype='int32')
The first problem I faced is creating the Embedding layer. It takes an input of 2D shape (batch, input_length), while my input here is 3D (batch,max_num_sents_in_doc, max_num_words_in_sent, )
I tried flattening the input before the embedding layer then reshaping it after the embeddings are retrieved, but failed. I also tried using TimeDistributed around my Embedding layer but the masking wasn't passed properly to the next layers.
Ideally what I want to do is something like: 
sequence = Input(shape=(max_num_sents_in_doc, max_num_words_in_sent, ), dtype='int32')
embedded = Embedding(vocab_size, emb_dim, mask_zero=True)(sequence)
lstm1 = TimeDistributed(LSTM(lstm_dim, return_sequences=False))(embedded)
dropout1 = Dropout(dropout_prob)(lstm1)
lstm2 = LSTM(lstm_dim, return_sequences=False)(dropout1)
dropout2 = Dropout(dropout_prob)(lstm2)

Thanks

**Comment by unrealwill** [comment]
Hello,

Don't use masks : They are passed down to the loss function but for a LSTM2D" it will pose problem. 
Prepad with 0.
TimeDistributed(Embedding) should work.

**Comment by rmdort** [comment]
@Youmna-H I am using the following model for hierarchical RNN for sentiment analysis.

````
max_sentences = 1
MAX_SEQUENCE_LENGTH = 40
NUM_CLASSES = 2

# Encode each timestep
in_sentence = Input(shape=(MAX_SEQUENCE_LENGTH,), dtype='int64')
embedded_sentence = Embedding(nb_words + 1, EMBEDDING_DIM, weights=[word_embedding_matrix],
                     trainable = False)(in_sentence)
lstm_sentence = LSTM(hidden_dim_1)(embedded_sentence)
encoded_model = Model(in_sentence, lstm_sentence)

sequence_input = Input(shape=(max_sentences, MAX_SEQUENCE_LENGTH), dtype='int64')
seq_encoded = TimeDistributed(encoded_model)(sequence_input)
seq_encoded = Dropout(0.2)(seq_encoded)

# Encode entire sentence
seq_encoded = LSTM(hidden_dim_2)(seq_encoded)

# Prediction
prediction = Dense(NUM_CLASSES, activation='softmax')(seq_encoded)
model = Model(sequence_input, prediction)
model.compile(loss='categorical_crossentropy',
              optimizer='rmsprop',
              metrics=['accuracy'])

print (model.summary())
````

Hope this helps

**Comment by erickrf** [comment]
@rmdort in this case, are you sure the second model backpropagates the error to the first one?
